### PR TITLE
Render jobs from templates

### PR DESCRIPTION
**Pull request by sjanota**
**Description**

The idea behind this PR is that a person doing a release needs only change a list of supported releases and all jobs will be generated automatically this means that "Define release jobs" and "Remove previous release jobs" from [release process](https://kyma-project.io/community/guidelines/release/#release-process-release-process) can be skipped.

How it would look like now:
1. Go to `templates/config.yaml` and change `global.releases` to support the last three releases.
2. Run `go run development/tools/cmd/rendertemplates/main.go --config templates/config.yaml`
3. Run `go test development/tools/jobs/...`. If anything is red fix it using these guidelines: 
  * For release tests using `GetKymaReleasesSince` with a release that is no longer supported change it to use `GetAllKymaReleases`.
  * For release tests using `GetKymaReleasesUntil` with a release that is no longer supported simply remove that part of the test.
4. When tests are green commit all jobs and you just got yourself new release jobs.

How it works:
1. There is a new tool called `rendertemplates`. 
2. In directory `templates` there is a `config.yaml`. This file describes what files to render from which templates and the variables to use.
3. In `templates/templates` are templates to use for rendering. 
4. Before release, no changes should be needed in any other file then `config.yaml`. You just new release, remove the oldest one and rerender jobs. 
5. In the sunny day scenario, you don't even have to touch tests, as a file with test constants is also generated. Sometimes you have to remove some test cases If they were using releases that got out of scope.

**Info for reviewers**
Some job configuration files were renamed, most of them refactored, thus the amount of code lines changes is overwhelming. Tests should verify if jobs remain semantically identical, so pay close attention to the tests instead of jobs themselves. Tests weren't changed much, so it should be fairly easy.

The second thing is the release process. I aimed to remove most manual work in test-infra repository. Basically, there should be no need to update any file other than `config.yaml` and version file, before branching out. Pay close attention if that is true and if I correctly created all the templates.

**Related issue(s)**

### CODE REVIEW

**Comment by sjanota** [comment]
Based on the changes in this PR I was able to further simplify tests: https://github.com/sjanota/test-infra/pull/3

File: development/tools/jobs/tester/releases.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package tester
+
+// DO NOT EDIT. THIS FILE IS GENERATED
```
Comment (crabtree) [inline development/tools/jobs/tester/releases.go]: As this file is generated, I'd rather remove the next comment, it is a bit misleading.

File: development/tools/jobs/common/common_test.go [hunk 2]
```diff
@@ -10,17 +10,17 @@ import (
...
 	// WHEN
-	for _, currentRelease := range tester.GetAllKymaReleaseBranches() {
-		t.Run(currentRelease, func(t *testing.T) {
+	for _, currentRelease := range tester.GetAllKymaReleases() {
+		t.Run(currentRelease.String(), func(t *testing.T) {
 			jobConfig, err := tester.ReadJobConfig("./../../../../prow/jobs/kyma/common/common.yaml")
 			// THEN
 			require.NoError(t, err)
-			actualPresubmit := tester.FindPresubmitJobByName(jobConfig.Presubmits["kyma-project/kyma"], tester.GetReleaseJobName("kyma-common", currentRelease), currentRelease)
+			actualPresubmit := tester.FindPresubmitJobByName(jobConfig.Presubmits["kyma-project/kyma"], tester.GetReleaseJobName("kyma-common", currentRelease), currentRelease.Branch())
```
[the remark this answers is not in the document]
Comment (sjanota) [reply development/tools/jobs/common/common_test.go]: I already have a PR to do that: https://github.com/sjanota/test-infra/pull/3. It's not in the same PR as here I heavily rely on tests to catch any semantic differences.

**Comment by michal-hudy** [comment]
How updates are handled? For example https://github.com/kyma-project/test-infra/pull/1396. So for old releases, I need a different configuration than the current version.

File: docs/prow/prow-jobs.md [hunk 3]
```diff
@@ -81,6 +81,8 @@ If you want to trigger your job again, add one of these comments to your PR:
 
 ## Create jobs
 
+Jobs should be generated from templates in `templates` directory. For details on how to do that see [Configure templates](./templates.md).
```
Comment (majakurcius) [inline docs/prow/prow-jobs.md]:
```suggestion
Generate jobs from templates in the `templates` directory. For details on how to configure templates, see [Configure templates](./templates.md).
```

File: docs/prow/templates.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# Templates
+
+Jobs and Prow configuration are generated from templates that are in `templates` directory. There are following items there:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
Jobs and Prow configuration are generated from the templates in the `templates` directory. The directory contains these items:
```

File: docs/prow/templates.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+- `config.yaml` - config file that describes files to generate
+- `templates` - directory to gather all template files
+
+### Config structure
```
Comment (majakurcius) [inline docs/prow/templates.md]: Should we use the full word instead of `config`?

File: docs/prow/templates.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+Config file is expected to have two keys:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
The config file has two keys:
```
Is expected or has?

File: docs/prow/templates.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# Templates
+
+Jobs and Prow configuration are generated from templates that are in `templates` directory. There are following items there:
+
+- `config.yaml` - config file that describes files to generate
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `config.yaml` is the config file that describes files to generate.
```

File: docs/prow/templates.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+- `templates` - directory to gather all template files
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `templates` is the directory to gather all the template files.
```

File: docs/prow/templates.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+### Config structure
+
+Config file is expected to have two keys:
+
+- `global` - a map of values available for all templates
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `global` is a map of values available for all templates.
```

File: docs/prow/templates.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+- `templates` - a list of files to generate
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `templates` is a list of files to generate.
```

File: docs/prow/templates.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+Complete structure can be found [here](../../development/tools/cmd/rendertemplates/main.go).
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
To see the complete structure, go [here](../../development/tools/cmd/rendertemplates/main.go).
```

File: docs/prow/templates.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+Other then that `config.yaml` uses YAML techniques for better organization, so normal YAML rules apply.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
Other then that, `config.yaml` uses YAML techniques for better organization, so normal YAML rules apply.
```
This sounds somewhat too colloquial... 
Maybe change it to something like:
> `config.yaml` uses YAML techniques for better organization. If you use YAML, you are already familiar with the rules. 

?

File: docs/prow/templates.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+### Template development
+
+A template will receive two objects as inputs:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
A template receives two objects as input:
```
Don't use future tense (_will_), use present tense where possible.

File: docs/prow/templates.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+- `Values` which contains all `values` specified in its configuration in config file
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `Values`, which contains all the values specified under `values` in the configuration file
```

File: docs/prow/templates.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+- `Global` which contains values specified under `global` in config file
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `Global` which contains values specified under `global` in the configuration file.
```

File: docs/prow/templates.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+All functions from [`sprig`](https://github.com/Masterminds/sprig) library are available in templates. It is the same library that is used by helm, so if you know helm there's nothing new. Also some additional functions are available:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
All the functions from [`sprig`](https://github.com/Masterminds/sprig) library are available in the templates. It is the same library that is used by helm, so if you know helm, you are already familiar with them. Also, a few additional functions are available:
```

Also, do we capitalize `helm`?

File: docs/prow/templates.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+- `releaseMatches <release> <since> <until>` - returns a bool value indicating if `release` fits in the range. You can use `nil` to remove one of the bounds. Example: `releaseMatches {{ $rel }} '1.2' '1.5'` will check if release `$rel` is younger then `1.2` and older then `1.5`.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `releaseMatches <release> <since> <until>` returns a boolean value indicating whether `release` fits in the range. Use `nil` to remove one of the bounds. For example: `releaseMatches {{ $rel }} '1.2' '1.5'` checks if the release `$rel` is younger than `1.2` and older than `1.5`.
```

File: docs/prow/templates.md [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+- `matchingReleases <all-releases> <since> <until>` - returns a filtered list of releasesto only those that fits in the range.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `matchingReleases <all-releases> <since> <until>` returns a list of releases filtered to only those that fit in the range.
```

File: docs/prow/templates.md [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+### Add new component
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
### Add a new component
```

File: docs/prow/templates.md [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+To add new component find a `templates` entry for `templates/component.yaml`. Then add new entry with your component to the `render` list. An example below defines a component in Kyma repository using go1.12 buildpack:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
To add a new component, find a `templates` entry for `templates/component.yaml`. Then, add a new entry with your component to the `render` list.  
This example defines a component in the Kyma repository using go1.12 buildpack:
```

File: docs/prow/templates.md [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  - from: templates/component.yaml
+    render:
+      - to: ../prow/jobs/kyma/components/new-component/new-component.yaml
+        values:
+          <<: *go_kyma_component_1_12
+          path: components/new-component
+    ...
+```
+
+If buildpack you're willing to use is not there yet you have to add it. The best would be to follow existing buildpacks.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
If the buildpack you want to use is not there yet, add it. If possible, use existing buildpacks.
```
- It would be best to inform the user why they should refrain from using other buildpacks. If only to avoid the hassle, consider removing the second sentence. 
- `You're willing to` has a slightly different meaning. See [here](https://dictionary.cambridge.org/dictionary/english/willing).

File: docs/prow/templates.md [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+When writing tests for new component use function `tester.GetKymaReleasesSince(<next release>)` to create release jobs tests. This will automatically check if new release jobs where created when doing release.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
When writing tests for the new component, use the function `tester.GetKymaReleasesSince(<next release>)` to create release jobs tests. This automatically checks whether new release jobs were created when doing release.
```

File: docs/prow/templates.md [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+
+### Change component job configuration
+
+Whenever you need to change component job configuration follow this steps:
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
To change component job configuration, follow these steps:
```
Add the appropriate article (`a`/`the`) before `component job configuration`.

File: docs/prow/templates.md [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+1. In `config.yaml` change the name of the file where the jobs are generated. A suffix like `deprecated` should be enough. Change this path also in tests.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
1. In `config.yaml`, change the name of the file where the jobs are generated. For example, add the suffix `deprecated`. Change this path in tests as well.
```

- Please don't speculate. Don't use `should`. Either it is enough or it's not. 
- Avoid using passive voice when possible. Can you rephrase `file where the jobs are generated` to something like:
`file containing the generated jobs`?

File: docs/prow/templates.md (hunk 6 above)
Comment (sjanota) [reply docs/prow/templates.md]: Currently it has, but at the same time it has to :D

File: docs/prow/templates.md (hunk 21 above)
Comment (sjanota) [reply docs/prow/templates.md]: I'd rather meant to suggest following the style in which other buildpacks are defined. I'll try to rephrase it.

File: docs/prow/prow-jobs.md (hunk 3 above)
Comment (sjanota) [reply docs/prow/prow-jobs.md]: I resolved it differnetly.

File: docs/prow/templates.md (hunk 12 above)
Comment (sjanota) [reply docs/prow/templates.md]: On the second thought, if you see it is a YAML you may expect all the YAML things to be there, so I removed the sentence completely.

File: docs/prow/templates.md [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+
+### Change component job configuration
+
+To change component job configuration, follow these steps:
+1. In `config.yaml`, change the name of the file where the jobs are generated. For example, add the suffix `deprecated`. Change this path in tests as well.
+2. Add `until: <last release>` to this configuration.
+3. Create new entry with new configuration. It should generate new jobs to the file used before.
+4. Add `since: <next release>` to new entry.
+
+Example: buildpack for api-controller has changed from go1.11 to go.12 in release 1.5. Before the change component configuration looked like this:
```
Comment (tomekpapiernik) [inline docs/prow/templates.md]:
```suggestion
Example: buildpack for the API Controller has changed from `go1.11` to `go.12` in release 1.5. This is the component configuration before the buildpack change:
```

File: docs/prow/templates.md [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+```yaml
+      - to: ../prow/jobs/kyma/components/api-controller/api-controller.yaml
+        values:
+          <<: *go_kyma_component_1_11
+          path: components/api-controller
+```
+
+New configuration will look like follows:
```
Comment (tomekpapiernik) [inline docs/prow/templates.md]:
```suggestion
This is what the configuration created after the buildpack change looks like:
```

File: docs/prow/templates.md [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+If the buildpack you want to use is not there yet you have to add it. When you add a new buildpack follow the example of already defined ones.
+
+When writing tests for a new component, use the `tester.GetKymaReleasesSince(<next release>)` function to create release jobs tests. This automatically checks whether new release jobs were created in the release process.
+
+### Change component job configuration
+
+To change component job configuration, follow these steps:
+1. In the `config.yaml` file, change the name of the file where the jobs are generated. For example, add the suffix `deprecated`. Change the path to this file in tests accordingly.
+2. Add `until: <last release>` to this configuration.
+3. Create a new entry with a new configuration. It should generate new jobs to the file used before.
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
3. Create a new entry with the new configuration. Set the `to` field to point to the file responsible for storing jobs.
```

File: docs/prow/templates.md [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+
+- `global` is a map of values available for all templates.
+- `templates` is a list of files to generate.
+
+To see the complete structure, see [this](../../development/tools/cmd/rendertemplates/main.go) file.
+
+### Template development
+
+A template receives two objects as input:
+- `Values`, which contains all the values specified under `values` in the configuration file
```
Comment (majakurcius) [inline docs/prow/templates.md]:
```suggestion
- `Values`, which contains all the values specified under `values` in the configuration file.
```
